ast/compile_test.go

```diff
@@ -2335,6 +2335,284 @@ elsekw {
+			note: "self imported and accessed, no metadata",
+			module: `package test
+import future.self
+
+p {
+	self.metadata.chain()[0].path == ["test", "p"]
+	self.metadata.rule() == {}
+}`,
+			exp: `package test
+
+p = true { 
+	__local2__ = [{"path": ["test", "p"]}]
+	__local3__ = {}
+	__local0__ = __local2__
+	equal(__local0__[0].path, ["test", "p"])
+	__local1__ = __local3__
+	equal(__local1__, {}) 
+}`,
+		},
+		{
+			note: "self imported and accessed, with metadata",
+			module: `# METADATA
+# description: A test package
+package test
+
+import future.self
+
+# METADATA
+# title: My P Rule
+p {
+	self.metadata.chain()[0].title == "My P Rule"
+	self.metadata.chain()[1].description == "A test package"
+}
+
+# METADATA
+# title: My Other P Rule
+p {
+	self.metadata.rule().title == "My Other P Rule"
+}`,
+			exp: `# METADATA
+# {"scope":"package","description":"A test package"}
+package test
+
+# METADATA
+# {"scope":"rule","title":"My P Rule"}
+p = true { 
+	__local3__ = [
+		{"annotations": {"scope": "rule", "title": "My P Rule"}, "path": ["test", "p"]}, 
+		{"annotations": {"description": "A test package", "scope": "package"}, "path": ["test"]}
+	]
+	__local0__ = __local3__
+	equal(__local0__[0].title, "My P Rule")
+	__local1__ = __local3__
+	equal(__local1__[1].description, "A test package") 
+}
+
+# METADATA
+# {"scope":"rule","title":"My Other P Rule"}
+p = true { 
+	__local4__ = {"scope": "rule", "title": "My Other P Rule"}
+	__local2__ = __local4__
+	equal(__local2__.title, "My Other P Rule")
+}`,
+		},
+		{
+			note: "self referenced multiple times",
+			module: `# METADATA
+# description: TEST
+package test
+
+import future.self
+
+p {
+	self.metadata.chain()[0].path == ["test", "p"]
+	self.metadata.chain()[1].path == ["test"]
+}`,
+			exp: `# METADATA
+# {"scope":"package","description":"TEST"}
+package test
+
+p = true { 
+	__local2__ = [
+		{"path": ["test", "p"]}, 
+		{"annotations": {"description": "TEST", "scope": "package"}, "path": ["test"]}
+	]
+	__local0__ = __local2__
+	equal(__local0__[0].path, ["test", "p"])
+	__local1__ = __local2__
+	equal(__local1__[1].path, ["test"]) }`,
+		},
+		{
+			note: "self return value",
+			module: `package test
+import future.self
+
+p := self.metadata.chain()`,
+			exp: `package test
+
+p := __local0__ { 
+	__local1__ = [{"path": ["test", "p"]}]
+	true
+	__local0__ = __local1__ 
+}`,
+		},
+		{
+			note: "self argument in function call",
+			module: `package test
+import future.self
+
+p {
+	q(self.metadata.chain())
+}
+
+q(s) {
+	s == ["test", "p"]
+}`,
+			exp: `package test
+
+p = true { 
+	__local2__ = [{"path": ["test", "p"]}]
+	__local1__ = __local2__
+	data.test.q(__local1__) 
+}
+
+q(__local0__) = true { 
+	equal(__local0__, ["test", "p"]) 
+}`,
+		},
+		{
+			note: "self used in array comprehension",
+			module: `package test
+import future.self
+
+p = [x | x := self.metadata.chain()]`,
+			exp: `package test
+
+p = [__local0__ | __local1__ = __local2__; __local0__ = __local1__] { 
```

Comment:
The chosen strategy of replacing all calls to `self.metadata.*` functions with an assignment to the output var shows its ugly face in particular here.
An alternative would be to replace all uses of the original var with the injected metadata var. But that would be more complex, and I don't know if that would produce more edge cases (and bugs).